### PR TITLE
Use new journal

### LINKED ISSUE #6307: Replace old journal with new io.zeebe.journal.Journal API
**Description**

As part of #6147, and after #6304, we now should replace the old journal with the new API. As a first step, we will keep the same `RaftLog` API where applicable, and simply adapt it for the new underlying journal. However, some changes may need to be made - if these are too big, please split into further issues.

Changes: 
1. Drop `getCompactableIndex` and `isCompactable` - this is not necessary any more, and we can simply just call `journal.deleteBefore` all the time. We should move the logging of whether compaction occurs or is skipped into the journal.
1. Reimplement `RaftLogReader` to use the new Journal API.

Additionally, keep the `JournalReader.MODE` for now - the next step after this issue will remove it and replace it with different interfaces for each mode, but let's keep this PR small for now.

Blocked by #6304

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR refactors the RaftLog to use the new `zeebe-journal` module instead of the one in `atomix-storage`. It mostly rewrites `RaftLog` and `RaftLogReader`, and modifies the consumers of each.

To try to keep noise to a minimum, I've kept the names of the methods as they were - reset, truncate, compact, etc. This will need to be aligned to the journal API later. For new methods I've already aligned them with the new API (e.g. seekToAsqn). For the same reason, the old journal is still present, and should be deleted in a subsequent PR. Similarly, some configuration options are still present but unused - this should be addressed in a separate PR to change how configuration is done.

## Related issues

closes #6307 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: journal/src/main/java/io/zeebe/journal/file/SparseJournalIndex.java [hunk 1]
```diff
@@ -30,6 +30,7 @@
 
   public SparseJournalIndex(final int density) {
     this.density = density;
+    assert density > 0;
```
Comment (npepinpe) [inline journal/src/main/java/io/zeebe/journal/file/SparseJournalIndex.java:33]: Oops, this was for debugging.

File: logstreams/src/main/java/io/zeebe/logstreams/storage/atomix/AtomixLogStorageReader.java [hunk 2]
```diff
@@ -50,11 +49,14 @@ public boolean isEmpty() {
 
   @Override
   public long read(final DirectBuffer readBuffer, final long address) {
-    if (address < reader.getFirstIndex()) {
+    final long index = reader.reset(address);
+
+    // TODO: check if important to return OP_RESULT_INVALID_ADDR if we tried to seek before
```
Comment (npepinpe) [inline logstreams/src/main/java/io/zeebe/logstreams/storage/atomix/AtomixLogStorageReader.java:54]: Solved, can be removed

**Comment by npepinpe** [comment]
Notable changes:

- Remove isCompactable/getCompactable index. All logic of compaction is pushed down to the journal, and we just always call compact regardless if anything will get compacted or not.
- Removed ZeebeIndexAdapter and related stuff to rely entirely on ASQN
- Only `ZeebeEntry` entries have an ASQN, and we use the `getLowestPosition` as the ASQN. You will see several times checks if an entry has type `ZeebeEntry.class`, which is pretty hacky and should be improved later by introducing the concept of an application entry (or batched application entry) which always defines an ASQN.
- `RaftLog` currently tracks the last appended entry since we need to know its index and term as part of the Raft protocol. This is hacky and should be improved, ideally we don't want to keep track of that in memory, we just want a way to get the term/index of the tip of the log.
- The commit index is now tracked in both the `RaftLog` and the `RaftContext`. Ideally we either track it as a volatile in one place, OR we have the central update location (e.g. `RaftContext#setCommitIndex` or `RaftLog#setCommitIndex`) propagate to listeners in preparation for us migrating to actors/thread-safe model.
- `RaftLogReader#seekToLast` seems unused, but I kept it for now as it's aligned with the journal reader. We should clean up at the end and remove anything unused.
- When appending, the `RaftLog` will check and cast the entry if it's a `ZeebeEntry` and then append with the ASQN. This is also hacky, and ideally we find a better way to append stuff that doesn't rely on this. Maybe we only have `Journal#append(asqn, data)` but we can pass `ASQN_IGNORE`?
- `RaftLog#append(entry, checksum)` currently ignores the checksum - this is because we cannot construct a journal record since we don't know the expected index yet. This should be fixed when we change the serialization and the whole `JournalRecord` is replicated.
- I've kept the `flushExplicitly` flag for compatibility.
- `RaftLogReader` still uses the `Mode` enum - this should be replaced by a `RaftLogReader` interface, and a separate implementation for each mode.
- `RaftLogReader` keeps track of the next index to return the correct value in `hasNext()` if it's in commit mode. This is probably fine as in commit mode the next index is either there or compacted, but never "deleted", so even if it's not kept up to date, it shouldn't cause issues. Still, we could consider either a `getCurrentEntry` again or a `peek()` call.
- `RaftLogReader#seekToAsqn` will throw an `UnsupportedOperationException` at the moment if we try to seek to an uncommitted ASQN. This is because otherwise we would need to walk the log backwards (which isn't supported except via hacky workarounds), and find the latest ASQN. This is the second place which would benefit from a `seekToLastASQN` call (the first being the `LeaderRole` which needs the latest entry with an ASQN so it can compare later appends).
- `AtomixLogDeletionTest` has too much knowledge about segments and should be rewritten. In fact, anything in `logstreams` which knows about segments should be rewritten - this is completely implementation detail.
- `AtomixRecordEntrySupplier` is possibly not needed anymore and we can maybe just use a reader directly.
- Nothing is documented for now (`RaftLog` or `RaftLogReader`) as I assume the interfaces aren't fixed yet - we should do that after.
- Consider introducing the concept of `ApplicationEntry` and `BatchedApplicationEntry` and push the validation of ASQN down to the journal - we can simply be more strict and say ASQN must be > 0, and each ASQN must be 1 greater than the previous one, without gaps (as we do with the log sequence number LSQN).
- `AtomixLogStorageReader` is a good candidate to use `seekToFirstASQN()` and `seekToLastASQN()` (or support `seekToAsqn(Long.MIN_VALUE)` and `seekToAsqn(Long.MAX_VALUE)`). We want to seek to the last one on recovery to update the dispatcher position, and the `LogStreamReader` typically wants to seek to the first block as a means to reset its state (`lookUpApproximateAddress(Long.MIN_VALUE)`). Right now that last case we handle by checking if the given position is `Long.MIN_VALUE`, which is hacky, and then seek to the beginning and find the first ASQN-valid entry.
- `AtomixLogStorageReader` now adds an additional seek when reading - when doing a `LogStreamReader#seek`, it will first `lookUpApproximateAddress` (which seeks possibly multiple times and tries to find the entry with the closest position), then will do a `read`, which will again seek to the index (which is now the previous index, as we had to read the entry to know it was the right one). If we were already behind it anyway, `JournalReader#seek` to the same index is not a NOOP right now (though it should be).
- No new tests were written for `RaftLog` and `RaftLogReader`, as again, I imagine the implementations are not anywhere near finished. This should also be done once they start to stabilize.

**Comment by npepinpe** [comment]
Collecting all the TODOs in the code before removing them:

- LeaderRole#findLastZeebeEntry: `TODO: good candidate to use a seekToLastASQN()`
- AtomixRecordEntrySupplier: consider getting rid of this now that the ASQN stuff is supported directly through the log
- AtomixLogDeletionServiceTest: rewrite these tests to not know implementation details of compaction

RaftLog:

 - get rid of keeping track of the last appended entry
 - align names with journal API (e.g. compact, truncate, etc.)
 - simplify appending logic, remove hack to detect if there's an ASQN, etc.
 - consider getting rid of RaftLogReader knowing about RaftLog, and instead have event listeners on the log which can be aware of new commits, new entries, truncation/compaction, etc. This may help us move forward with the whole thread-safe reader concept (and further down the line with having actors, e.g. RaftLog actor, RaftLogReader actor, etc., where they don't share memory but receive messages when things happen)
 - consider introducing an ApplicationEntry (instead of ZeebeEntry) which ties in with the concept of ASQN
 - consider introducing a BatchedApplicationEntry which contains multiple ApplicationEntry (or just the data, not the whole framing), such that we could do the whole validation on ASQNs based on this and not on anything Zeebe specific. This means adding the constraint that ASQN is also a function which starts at some X >= 0, and increases by one with no gaps.
- clean up unused methods
- consider getting rid of this and have consumers keep track of the last entry as it's easy to forget to reset it when modifying the log the usage of getLastEntry() is purely to get the index/term of the last entry, we never actually look at the data
- remove usage of builder in favour of IoC and configuration classes, where the config can be passed along instead of having all the parameters in all these builders


RaftLogReader

- get rid of keeping track of the next index; maybe some kind of `journalReader#peek()`?
- get rid of the Mode, and instead extract an interface and implement two different readers
- remove direct dependency on the RaftLog, instead pass messages between (RaftLogEventListener?)
- consider removing `seekToLast` from `RaftLogReader` as it's not used at all
- consider using bounding `seekToAsqn`, and have it support cases with `Long.MIN` or `Long.MAX` (or just `seekToFirstASQN` and `seekToLastASQN`)

AtomixLogStorageReader

- could use `seekToLastASQN`, `seekToFirstASQN`
- would benefit from having a `peek` call so we can tell if the next entry is the right position, or a way to know if seekToAsqn points to an application entry or not (does not need to be the exact position requested)
- probably needs to rewrite how the log stream reads in general
- make RaftLogReader#seek to the same index twice a NOOP
- avoid `lookUpApproximateAddress` having to find the entry (so seeking right past it so we read it), then `read` seeking again to it.

**Comment by npepinpe** [comment]
I'm running a load tests right now as `np-6307-new-journal`. I expect the performance will go dooooown :smile:

**Comment by deepthidevaki** [comment]
The performance of the weekly benchmark is also very low. Could that be related?

**Comment by Zelldon** [comment]
How good that we have that :laughing:

**Comment by Zelldon** [comment]
Maybe something we did with the engine migration (@saig0 ? ) or job workers again ? @korthout

**Comment by npepinpe** [comment]
My suspicion is getting rid of the getCurrentEntry optimization in atomixlogstorage reader, which causes every read to be a seek and not a next

**Comment by Zelldon** [comment]
And this was merged already?

**Comment by Zelldon** [comment]
If so we could build a benchmark without that merge

File: logstreams/src/main/java/io/zeebe/logstreams/storage/atomix/AtomixLogStorageReader.java [hunk 3]
```diff
@@ -50,11 +49,13 @@ public boolean isEmpty() {
 
   @Override
   public long read(final DirectBuffer readBuffer, final long address) {
-    if (address < reader.getFirstIndex()) {
+    final long index = reader.reset(address);
```
Comment (deepthidevaki) [inline logstreams/src/main/java/io/zeebe/logstreams/storage/atomix/AtomixLogStorageReader.java:52]: This reset can also be avoided by checking `reader.getNextEntry`

Comment (npepinpe) [reply logstreams/src/main/java/io/zeebe/logstreams/storage/atomix/AtomixLogStorageReader.java:52]: And that was the big one? That's...unexpected :sweat_smile:

**Comment by korthout** [comment]
> Maybe something we did with the engine migration (@saig0 ? ) or job workers ? @korthout

I don't think the new job workers are used when you start a new benchmark.

I also don't think the engine will cause any problems at this time. The new bpmn element commands have not yet been introduced. Most of the logic is still the same.

**Comment by npepinpe** [comment]
With the latest changes it's at 90 WI/s, which is a drop of ~20-30 WI/s if you compare to other single partition benchmarks. This could be explained by double serialization, but also might be explained by other factors such as engine changes, since we saw drops even before this branch - I'm trying to find the exact commit so we can more properly compare.

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogReader.java [hunk 4]
```diff
@@ -18,50 +18,122 @@
...
+      final RaftLog log, final JournalReader journalReader, final RaftLogReader.Mode mode) {
+    this.log = log;
+    this.journalReader = journalReader;
+    this.mode = mode;
   }
 
   @Override
   public boolean hasNext() {
-    return delegate.hasNext();
+    return (mode == Mode.ALL || nextIndex <= log.getCommitIndex()) && journalReader.hasNext();
```
Comment (deepthidevaki) [inline atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogReader.java:46]: `nextIndex` is not initialized. Could that be a problem? If I check the usage, immediately after creating the reader, it is reset to a specific index. So I guess it is ok.

File: broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java [hunk 5]
```diff
@@ -133,10 +131,8 @@ private void createSnapshot(final long index) {
   private static RaftStorage.Builder builder(
       final RaftStorage.Builder builder, final TemporaryFolder folder) {
     try {
-      final int length = getEntrySize();
       return builder
-          // hardcode max segment size to allow a single entry only
-          .withMaxSegmentSize(JournalSegmentDescriptor.BYTES + 2 * Integer.BYTES + length)
+          .withMaxSegmentSize(JournalSegmentDescriptor.BYTES + 182)
```
Comment (deepthidevaki) [inline broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java:135]: What is this magic value 182?

File: journal/src/main/java/io/zeebe/journal/JournalReader.java [hunk 6]
```diff
@@ -90,4 +90,7 @@
    * @return the index of the record that will be returned by {@link #next()}
    */
   long seekToAsqn(long asqn);
+
+  @Override
+  void close();
```
Comment (deepthidevaki) [inline journal/src/main/java/io/zeebe/journal/JournalReader.java:95]: Why is this needed? The interface extends `AutoCloseable`.

File: journal/src/main/java/io/zeebe/journal/file/SegmentedJournal.java [hunk 7]
```diff
@@ -147,7 +160,7 @@ public long getFirstIndex() {
     if (!isEmpty()) {
       return segments.firstEntry().getValue().index();
     } else {
-      return 0;
+      return 1;
```
Comment (deepthidevaki) [inline journal/src/main/java/io/zeebe/journal/file/SegmentedJournal.java:163]: Why? If it is empty, should we return an index for which no entry exists? Returning 0 sounds correct to me.

File: logstreams/src/main/java/io/zeebe/logstreams/storage/atomix/AtomixLogStorageReader.java [hunk 8]
```diff
@@ -75,31 +66,17 @@ public long read(final DirectBuffer readBuffer, final long address) {
...
   public long readLastBlock(final DirectBuffer readBuffer) {
-    final var firstIndex = reader.getFirstIndex();
-    var index = reader.getLastIndex();
-
-    do {
-      reader.reset(index);
-      if (!reader.hasNext()) {
-        break;
-      }
+    reader.seekToAsqn(Long.MAX_VALUE);
```
Comment (deepthidevaki) [inline logstreams/src/main/java/io/zeebe/logstreams/storage/atomix/AtomixLogStorageReader.java:71]: Optional: Catch the exception incase we seek to non-committed. And use the previous logic of reading backwards as backup.

File: logstreams/src/main/java/io/zeebe/logstreams/storage/atomix/AtomixLogStorageReader.java [hunk 9]
```diff
@@ -111,50 +88,48 @@ public long readLastBlock(final DirectBuffer readBuffer) {
...
    *
    * @param index index to seek to
    */
   public Optional<Indexed<ZeebeEntry>> findEntry(final long index) {
-    // in the future, reset/seek to the same index will be a NOOP so we can just call it all the
-    // time; right now it's a bit slow, but we will immediately implement the new journal so this
-    // is fine
-    reader.reset(index);
+    if (reader.getNextIndex() != index) {
+      final long nextIndex = reader.reset(index);
```
Comment (deepthidevaki) [inline logstreams/src/main/java/io/zeebe/logstreams/storage/atomix/AtomixLogStorageReader.java:127]: RaftLogReader#reset also have checks nextIndex before reset. So we can remove it from here.

**Review by deepthidevaki — CHANGES_REQUESTED** [review]

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogReader.java (hunk 4 above)
Comment (npepinpe) [reply atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogReader.java:46]: It will be 0 by default, which should be less than the commit index if we've committed anything ever, and equal to the commit index if we've never committed anything. So we should probably initialize it to the first index :+1:

File: broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java (hunk 5 above)
Comment (npepinpe) [reply broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java:135]: It's exactly the entry size now, which will change again when we change the serialization. Like I wrote below, `AtomixLogDeletionServiceTest` should be rewritten so it doesn't know about segments, serialization and so on. It has too much implementation knowledge.

File: journal/src/main/java/io/zeebe/journal/JournalReader.java (hunk 6 above)
Comment (npepinpe) [reply journal/src/main/java/io/zeebe/journal/JournalReader.java:95]: To avoid handling the exception - none of the consumers expect close to throw an exception. I could avoid this and make it bubble up, or we can handle the error. I would prefer the latter, but I would also propose to do it as a follow up.

File: journal/src/main/java/io/zeebe/journal/file/SegmentedJournal.java (hunk 7 above)
Comment (npepinpe) [reply journal/src/main/java/io/zeebe/journal/file/SegmentedJournal.java:163]: If I recall correctly, there's some behaviour in Atomix which expects the first index to always be 1, and the last index to be 0 in that case. But it shouldn't be hardcoded and rather we should return the index of the first segment's descriptor (which happens to be 1).

File: journal/src/main/java/io/zeebe/journal/JournalReader.java (hunk 6 above)
Comment (deepthidevaki) [reply journal/src/main/java/io/zeebe/journal/JournalReader.java:95]: :+1:

**Comment by deepthidevaki** [comment]
RandomizedTest is failing consistently. It seems commitIndex is never set.

**Comment by npepinpe** [comment]
Weird as the tests were previously passing, so it must be one of the changes I did in the last couple commits.

File: logstreams/src/main/java/io/zeebe/logstreams/storage/atomix/AtomixLogStorageReader.java (hunk 9 above)
Comment (npepinpe) [reply logstreams/src/main/java/io/zeebe/logstreams/storage/atomix/AtomixLogStorageReader.java:127]: Yes I added it after :+1:

**Comment by npepinpe** [comment]
Note that the latest changes now run with ~110 WI/s, so the performance is back to normal.

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogReader.java (hunk 4 above)
Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogReader.java:46]: This is what is causing RandomizedTest to fail. Initially both nextIndex and commitIndex = 0.  So then CommittedReader still returns true for hasNext(), if there is atleast one uncommitted entry.

File: broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java (hunk 5 above)
Comment (deepthidevaki) [reply broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java:135]: :+1:

File: journal/src/main/java/io/zeebe/journal/file/SegmentedJournal.java (hunk 7 above)
Comment (deepthidevaki) [reply journal/src/main/java/io/zeebe/journal/file/SegmentedJournal.java:163]: Then I would then first verify if the usage of #getFirstIndex is correct. For me it doesn't seem to be semantically correct to return a valid index if the journal is empty.

Comment (npepinpe) [reply journal/src/main/java/io/zeebe/journal/file/SegmentedJournal.java:163]: There's some assumptions in the Raft code around that as well - see

RaftContext
```java
private void verifySnapshotLogConsistent() {
    final long currentSnapshotIndex = getCurrentSnapshotIndex();
    if ((currentSnapshotIndex <= 0L && raftLog.getFirstIndex() != 1)
        || (currentSnapshotIndex > 0L
            && currentSnapshot.getIndex() + 1 < raftLog.getFirstIndex())) {
      // There is no snapshot, but the log has been compacted!
      throw new IllegalStateException(
          String.format(
              "Expected to find a snapshot at index >= log's first index %d, but found snapshot %d",
              raftLog.getFirstIndex(), currentSnapshotIndex));
    }
  }
```

Or when we initialize the `LeaderAppender` initial index (which is either the last index, or the first index when the log is empty):

```java
leaderIndex =
        raft.getLog().isEmpty() ? raft.getLog().getFirstIndex() : raft.getLog().getLastIndex() + 1;
```

If first index is 0, then the server doesn't start because later when we want to commit the initial entries (also in LeaderAppender):

```java
  public CompletableFuture<Long> appendEntries(final long index) {
    raft.checkThread();

    if (index == 0) {
      return appendEntries();
    }

    if (index <= raft.getCommitIndex()) {
      return CompletableFuture.completedFuture(index);
    }

    // If there are no other stateful servers in the cluster, immediately commit the index OR
    // If there are no other active members in the cluster, update the commit index and complete the
    // commit.
    // The updated commit index will be sent to passive/reserve members on heartbeats.
    if (raft.getCluster().getActiveMemberStates().isEmpty()) {
      final long previousCommitIndex = raft.getCommitIndex();
      raft.setCommitIndex(index);
      completeCommits(previousCommitIndex, index);
      return CompletableFuture.completedFuture(index);
    }

    // Only send entry-specific AppendRequests to active members of the cluster.
    return appendFutures.computeIfAbsent(
        index,
        i -> {
          for (final RaftMemberContext member : raft.getCluster().getActiveMemberStates()) {
            appendEntries(member);
          }
          return new CompletableFuture<>();
        });
  }
```

See the line where it says `index == 0`? When committing the initial entries, we pass the leaderIndex, in which case it does this special snowflake thing of sending a heartbeat :smile:

Comment (deepthidevaki) [reply journal/src/main/java/io/zeebe/journal/file/SegmentedJournal.java:163]: Ok. In the leaderAppender it is already checking if it is empty before reading getFirstIndex, so it should be ok.  For the other check `currentSnapshotIndex <= 0L && raftLog.getFirstIndex() != 1)`, it should also add a check if the journal is empty. We can do it another PR.
For this PR, instead of returning 1, it should return firstSegment.index().

File: logstreams/src/main/java/io/zeebe/logstreams/storage/atomix/AtomixLogStorageReader.java [hunk 10]
```diff
@@ -141,6 +139,36 @@ private long readFirstBlock() {
...
   }
 
+  private long seekToFirst() {
+    // looking for ASQN 0 is a trick which will skip any initial records with ASQN_IGNORE
+    // but stop as soon as we find an ASQN with 0 or greater (and we typically start at 1)
+    return findEntry(0).map(Indexed::index).orElse(LogStorage.OP_RESULT_INVALID_ADDR);
+  }
+
+  /** @deprecated should be dropped and handled properly by the RaftLogReader or the journal */
+  @Deprecated(since = "1.0.0")
```
Comment (deepthidevaki) [inline logstreams/src/main/java/io/zeebe/logstreams/storage/atomix/AtomixLogStorageReader.java:149]: Is it really necessary? It is a private method :grinning:

Comment (npepinpe) [reply logstreams/src/main/java/io/zeebe/logstreams/storage/atomix/AtomixLogStorageReader.java:149]: I wanted it to be explicit so we would know to remove it in the future, or question why we're still using it :smile:

**Review by deepthidevaki — APPROVED** [review]
Looks good to me :+1:
